### PR TITLE
Backend

### PR DESCRIPTION
**Pull request by paul-crowdbotics**
Unit test for RegistrationController.registerUserAccount(...)

### CODE REVIEW

File: cookiecutter.json [hunk 1]
```diff
@@ -8,19 +8,14 @@
...
     "project_package_name": "{{ cookiecutter.project_slug|lower|replace(' ', '')|replace('-', '') }}",
-    "entity_id_type": "String",
-    "has_h2_database": "y",
+    "database_option": ["h2", "mysql", "postgresql", "mongodb", "none"],
+    "entity_id_type": ["Long", "String"],
     "has_lombok": "n",
-    "has_mongodb_database": "n",
-    "has_mysql_database": "n",
-    "has_postgresql_database": "n",
-    "has_redis_database": "n",
```
Comment (driverdan) [inline cookiecutter.json:17]: Did you mean to remove Redis?

File: {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+	private String token;
+
+	@OneToOne(targetEntity = User.class, fetch = FetchType.EAGER)
+	@JoinColumn(nullable = false, name = "user_id")
+	private User user;
+
+	private Date expiryDate;
+
+	public PasswordResetToken(final String token) {
```
Comment (charathram) [inline {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java:47]: If you have lombok, do you still need to specify all these constructors?

File: cookiecutter.json [hunk 3]
```diff
@@ -8,19 +8,14 @@
...
     "has_spring_boot_devtools": "y",
     "has_spring_boot_jms": "n",
-    "has_spring_boot_ldap": "n",
-    "has_spring_boot_oauth2": "y",
     "has_spring_boot_rest": "n",
     "has_spring_boot_session": "n",
     "java_version": "1.8",
+    "spring_boot_authentication": ["basic", "jwt", "ldap", "oauth2"],
```
Comment (driverdan) [inline cookiecutter.json:19]: I don't know much about Spring but couldn't you have more than one of these? For options that can have multiple values they need to remain separate.

**Review by driverdan — CHANGES_REQUESTED** [review]
I can't provide much feedback on your Java but have a few comments on how options are handled.

File: cookiecutter.json (hunk 1 above)
Comment (paul-crowdbotics) [reply cookiecutter.json:17]: At this time, yes.

File: cookiecutter.json (hunk 3 above)
Comment (paul-crowdbotics) [reply cookiecutter.json:19]: I guess you could, but I don't see the point.

File: {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java (hunk 2 above)
Comment (paul-crowdbotics) [reply {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java:47]: No, they are there in the case where Lombok is not picked as an option.

Comment (charathram) [reply {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java:47]: In that case, shouldn't the `endif` on line 27 be at the end of the class file, and this section in an `else`?

Comment (charathram) [reply {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java:47]: In that case, shouldn't the `endif` on line 27 be at the end of the file, and this section in an `else`?

Comment (paul-crowdbotics) [reply {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java:47]: No, there are 3 sections:
1) common code - regardless of Lombok setting
2) code when Lombok == 'y'
3) code when Lombok == 'n'

Comment (charathram) [reply {{cookiecutter.snake_case_slug}}/backend/src/main/java/{{cookiecutter.top_level_domain_name}}/{{cookiecutter.second_level_domain_name}}/{{cookiecutter.project_package_name}}/model/registration/PasswordResetToken.java:47]: Got it. Makes sense.

File: cookiecutter.json (hunk 3 above)
Comment (driverdan) [reply cookiecutter.json:19]: I suspect we'll have to break these out into separate options again in the future but in the interest of getting this merged it's fine for now. I can think of use cases for multiple methods, for example JWT and/or OAuth2 for the API and basic for the website.

**Review by driverdan — DISMISSED** [review]
I've pushed up all the arrays to [the scaffold class PR](https://github.com/crowdbotics/crowdbotics-slack-app/pull/1064). If any of the defaults need to be changed let me know. You'll need to change the values in cookiecutter.json to the default values.

**Review by driverdan — APPROVED** [review]

**Pull request merged**